{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Author: Peter Riesebos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Library and file imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = pd.read_csv(\"/groups/umcg-fg/tmp04/projects/gut-bulk/ongoing/2024-02-07-GutPublicRNASeq/datasets/GWAS/GCST90292538.h.tsv.gz\", sep='\\t')\n",
    "sum = pd.read_csv(\"/groups/umcg-fg/tmp04/projects/gut-bulk/ongoing/2024-02-07-GutPublicRNASeq/datasets/combined/mbqtl_output_combined_exp_fixed/merged_topeffects_final.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data manipulation and inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df = gwas.copy()\n",
    "sum_df = sum.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25429"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df = gwas_df[gwas_df[\"p_value\"] < 5e-8]\n",
    "gwas_df.chromosome.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>FreqSE</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>hm_coordinate_conversion</th>\n",
       "      <th>hm_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4168</th>\n",
       "      <td>1</td>\n",
       "      <td>1215852</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0724</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.1935</td>\n",
       "      <td>2.250000e-09</td>\n",
       "      <td>1_1215852_T_C</td>\n",
       "      <td>rs3813201</td>\n",
       "      <td>0.0759</td>\n",
       "      <td>0.1312</td>\n",
       "      <td>0.3268</td>\n",
       "      <td>++-+++</td>\n",
       "      <td>58.9</td>\n",
       "      <td>12.164</td>\n",
       "      <td>5</td>\n",
       "      <td>0.03260</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4173</th>\n",
       "      <td>1</td>\n",
       "      <td>1216593</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0745</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.1913</td>\n",
       "      <td>8.490000e-10</td>\n",
       "      <td>1_1216593_G_C</td>\n",
       "      <td>rs75972122</td>\n",
       "      <td>0.0698</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.3102</td>\n",
       "      <td>++-+++</td>\n",
       "      <td>61.0</td>\n",
       "      <td>12.810</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02523</td>\n",
       "      <td>lo</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4230</th>\n",
       "      <td>1</td>\n",
       "      <td>1221049</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0778</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.2427</td>\n",
       "      <td>7.534000e-10</td>\n",
       "      <td>1_1221049_C_G</td>\n",
       "      <td>rs944369</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0.2366</td>\n",
       "      <td>0.2548</td>\n",
       "      <td>????++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.427</td>\n",
       "      <td>1</td>\n",
       "      <td>0.51360</td>\n",
       "      <td>lo</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4469</th>\n",
       "      <td>1</td>\n",
       "      <td>1238902</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0694</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>1.230000e-08</td>\n",
       "      <td>1_1238902_T_C</td>\n",
       "      <td>rs4970364</td>\n",
       "      <td>0.0234</td>\n",
       "      <td>0.2535</td>\n",
       "      <td>0.3029</td>\n",
       "      <td>????++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.234</td>\n",
       "      <td>1</td>\n",
       "      <td>0.62880</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4535</th>\n",
       "      <td>1</td>\n",
       "      <td>1246371</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0777</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.1713</td>\n",
       "      <td>6.517000e-09</td>\n",
       "      <td>1_1246371_T_C</td>\n",
       "      <td>rs6603783</td>\n",
       "      <td>0.0890</td>\n",
       "      <td>0.0954</td>\n",
       "      <td>0.3118</td>\n",
       "      <td>++++++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.280</td>\n",
       "      <td>5</td>\n",
       "      <td>0.80930</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21611131</th>\n",
       "      <td>22</td>\n",
       "      <td>50007275</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0631</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.5266</td>\n",
       "      <td>1.222000e-11</td>\n",
       "      <td>22_50007275_T_C</td>\n",
       "      <td>rs5771222</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>0.4679</td>\n",
       "      <td>0.5568</td>\n",
       "      <td>+?+-++</td>\n",
       "      <td>56.6</td>\n",
       "      <td>9.226</td>\n",
       "      <td>4</td>\n",
       "      <td>0.05568</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21611132</th>\n",
       "      <td>22</td>\n",
       "      <td>50007493</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0644</td>\n",
       "      <td>0.0102</td>\n",
       "      <td>0.5274</td>\n",
       "      <td>3.152000e-10</td>\n",
       "      <td>22_50007493_C_T</td>\n",
       "      <td>rs137859</td>\n",
       "      <td>0.0195</td>\n",
       "      <td>0.4883</td>\n",
       "      <td>0.5573</td>\n",
       "      <td>???-++</td>\n",
       "      <td>76.4</td>\n",
       "      <td>8.475</td>\n",
       "      <td>2</td>\n",
       "      <td>0.01444</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21611133</th>\n",
       "      <td>22</td>\n",
       "      <td>50007537</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1110</td>\n",
       "      <td>0.0191</td>\n",
       "      <td>0.4274</td>\n",
       "      <td>6.632000e-09</td>\n",
       "      <td>22_50007537_A_AT</td>\n",
       "      <td>rs66520758</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4274</td>\n",
       "      <td>0.4274</td>\n",
       "      <td>?????+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21611137</th>\n",
       "      <td>22</td>\n",
       "      <td>50008121</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0781</td>\n",
       "      <td>0.0143</td>\n",
       "      <td>0.5466</td>\n",
       "      <td>4.545000e-08</td>\n",
       "      <td>22_50008121_C_A</td>\n",
       "      <td>rs137862</td>\n",
       "      <td>0.0255</td>\n",
       "      <td>0.4937</td>\n",
       "      <td>0.5646</td>\n",
       "      <td>+?+-?+</td>\n",
       "      <td>62.2</td>\n",
       "      <td>7.946</td>\n",
       "      <td>3</td>\n",
       "      <td>0.04714</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21611145</th>\n",
       "      <td>22</td>\n",
       "      <td>50008559</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0633</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>0.5294</td>\n",
       "      <td>1.383000e-11</td>\n",
       "      <td>22_50008559_C_T</td>\n",
       "      <td>rs137864</td>\n",
       "      <td>0.0235</td>\n",
       "      <td>0.4842</td>\n",
       "      <td>0.5684</td>\n",
       "      <td>+?+-++</td>\n",
       "      <td>53.6</td>\n",
       "      <td>8.612</td>\n",
       "      <td>4</td>\n",
       "      <td>0.07156</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25427 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chromosome  base_pair_location effect_allele other_allele    beta  \\\n",
       "4168               1             1215852             C            T  0.0724   \n",
       "4173               1             1216593             C            G  0.0745   \n",
       "4230               1             1221049             G            C  0.0778   \n",
       "4469               1             1238902             C            T  0.0694   \n",
       "4535               1             1246371             C            T  0.0777   \n",
       "...              ...                 ...           ...          ...     ...   \n",
       "21611131          22            50007275             C            T  0.0631   \n",
       "21611132          22            50007493             T            C  0.0644   \n",
       "21611133          22            50007537            AT            A  0.1110   \n",
       "21611137          22            50008121             A            C  0.0781   \n",
       "21611145          22            50008559             T            C  0.0633   \n",
       "\n",
       "          standard_error  effect_allele_frequency       p_value  \\\n",
       "4168              0.0121                   0.1935  2.250000e-09   \n",
       "4173              0.0121                   0.1913  8.490000e-10   \n",
       "4230              0.0126                   0.2427  7.534000e-10   \n",
       "4469              0.0122                   0.2703  1.230000e-08   \n",
       "4535              0.0134                   0.1713  6.517000e-09   \n",
       "...                  ...                      ...           ...   \n",
       "21611131          0.0093                   0.5266  1.222000e-11   \n",
       "21611132          0.0102                   0.5274  3.152000e-10   \n",
       "21611133          0.0191                   0.4274  6.632000e-09   \n",
       "21611137          0.0143                   0.5466  4.545000e-08   \n",
       "21611145          0.0094                   0.5294  1.383000e-11   \n",
       "\n",
       "                variant_id        rsid  FreqSE  MinFreq  MaxFreq Direction  \\\n",
       "4168         1_1215852_T_C   rs3813201  0.0759   0.1312   0.3268    ++-+++   \n",
       "4173         1_1216593_G_C  rs75972122  0.0698   0.1342   0.3102    ++-+++   \n",
       "4230         1_1221049_C_G    rs944369  0.0086   0.2366   0.2548    ????++   \n",
       "4469         1_1238902_T_C   rs4970364  0.0234   0.2535   0.3029    ????++   \n",
       "4535         1_1246371_T_C   rs6603783  0.0890   0.0954   0.3118    ++++++   \n",
       "...                    ...         ...     ...      ...      ...       ...   \n",
       "21611131   22_50007275_T_C   rs5771222  0.0241   0.4679   0.5568    +?+-++   \n",
       "21611132   22_50007493_C_T    rs137859  0.0195   0.4883   0.5573    ???-++   \n",
       "21611133  22_50007537_A_AT  rs66520758  0.0000   0.4274   0.4274    ?????+   \n",
       "21611137   22_50008121_C_A    rs137862  0.0255   0.4937   0.5646    +?+-?+   \n",
       "21611145   22_50008559_C_T    rs137864  0.0235   0.4842   0.5684    +?+-++   \n",
       "\n",
       "          HetISq  HetChiSq  HetDf  HetPVal hm_coordinate_conversion  hm_code  \n",
       "4168        58.9    12.164      5  0.03260                       lo       10  \n",
       "4173        61.0    12.810      5  0.02523                       lo        5  \n",
       "4230         0.0     0.427      1  0.51360                       lo        5  \n",
       "4469         0.0     0.234      1  0.62880                       lo       10  \n",
       "4535         0.0     2.280      5  0.80930                       lo       10  \n",
       "...          ...       ...    ...      ...                      ...      ...  \n",
       "21611131    56.6     9.226      4  0.05568                       lo       10  \n",
       "21611132    76.4     8.475      2  0.01444                       lo       10  \n",
       "21611133     0.0     0.000      0  1.00000                       lo       10  \n",
       "21611137    62.2     7.946      3  0.04714                       lo       10  \n",
       "21611145    53.6     8.612      4  0.07156                       lo       10  \n",
       "\n",
       "[25427 rows x 20 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df = gwas_df.drop_duplicates(subset='rsid', keep='first')\n",
    "gwas_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7706"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df.Gene.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7620"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df = sum_df.drop_duplicates(subset='SNP', keep='first')\n",
    "sum_df.Gene.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3999925/3697496330.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_df[\"varbeta\"] = gwas_df[\"standard_error\"] ** 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>...</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>hm_coordinate_conversion</th>\n",
       "      <th>hm_code</th>\n",
       "      <th>varbeta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4168</th>\n",
       "      <td>1</td>\n",
       "      <td>1215852</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0724</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.1935</td>\n",
       "      <td>2.250000e-09</td>\n",
       "      <td>1_1215852_T_C</td>\n",
       "      <td>rs3813201</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1312</td>\n",
       "      <td>0.3268</td>\n",
       "      <td>++-+++</td>\n",
       "      <td>58.9</td>\n",
       "      <td>12.164</td>\n",
       "      <td>5</td>\n",
       "      <td>0.03260</td>\n",
       "      <td>lo</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4173</th>\n",
       "      <td>1</td>\n",
       "      <td>1216593</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0745</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0.1913</td>\n",
       "      <td>8.490000e-10</td>\n",
       "      <td>1_1216593_G_C</td>\n",
       "      <td>rs75972122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.3102</td>\n",
       "      <td>++-+++</td>\n",
       "      <td>61.0</td>\n",
       "      <td>12.810</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02523</td>\n",
       "      <td>lo</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4230</th>\n",
       "      <td>1</td>\n",
       "      <td>1221049</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0778</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>0.2427</td>\n",
       "      <td>7.534000e-10</td>\n",
       "      <td>1_1221049_C_G</td>\n",
       "      <td>rs944369</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2366</td>\n",
       "      <td>0.2548</td>\n",
       "      <td>????++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.427</td>\n",
       "      <td>1</td>\n",
       "      <td>0.51360</td>\n",
       "      <td>lo</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome  base_pair_location effect_allele other_allele    beta  \\\n",
       "4168           1             1215852             C            T  0.0724   \n",
       "4173           1             1216593             C            G  0.0745   \n",
       "4230           1             1221049             G            C  0.0778   \n",
       "\n",
       "      standard_error  effect_allele_frequency       p_value     variant_id  \\\n",
       "4168          0.0121                   0.1935  2.250000e-09  1_1215852_T_C   \n",
       "4173          0.0121                   0.1913  8.490000e-10  1_1216593_G_C   \n",
       "4230          0.0126                   0.2427  7.534000e-10  1_1221049_C_G   \n",
       "\n",
       "            rsid  ...  MinFreq  MaxFreq  Direction HetISq  HetChiSq  HetDf  \\\n",
       "4168   rs3813201  ...   0.1312   0.3268     ++-+++   58.9    12.164      5   \n",
       "4173  rs75972122  ...   0.1342   0.3102     ++-+++   61.0    12.810      5   \n",
       "4230    rs944369  ...   0.2366   0.2548     ????++    0.0     0.427      1   \n",
       "\n",
       "      HetPVal  hm_coordinate_conversion hm_code   varbeta  \n",
       "4168  0.03260                        lo      10  0.000146  \n",
       "4173  0.02523                        lo       5  0.000146  \n",
       "4230  0.51360                        lo       5  0.000159  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df[\"varbeta\"] = gwas_df[\"standard_error\"] ** 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneChr</th>\n",
       "      <th>GenePos</th>\n",
       "      <th>GeneStrand</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNPChr</th>\n",
       "      <th>SNPPos</th>\n",
       "      <th>SNPAlleles</th>\n",
       "      <th>SNPEffectAllele</th>\n",
       "      <th>...</th>\n",
       "      <th>DatasetZScores.1000IBD.ERP113396.ERP114636.GTEx_Colon_Transverse.GTEx_Small_Intestine.SRP063496.SRP068609.SRP077046.SRP096757.SRP129004.SRP155976.SRP189239.</th>\n",
       "      <th>DatasetSampleSizes.1000IBD.ERP113396.ERP114636.GTEx_Colon_Transverse.GTEx_Small_Intestine.SRP063496.SRP068609.SRP077046.SRP096757.SRP129004.SRP155976.SRP189239.</th>\n",
       "      <th>NrTestedSNPs</th>\n",
       "      <th>ProportionBetterPermPvals</th>\n",
       "      <th>BetaDistAlpha</th>\n",
       "      <th>BetaDistBeta</th>\n",
       "      <th>BetaAdjustedMetaP</th>\n",
       "      <th>PvalueNominalThreshold</th>\n",
       "      <th>qval</th>\n",
       "      <th>varbeta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000164308</td>\n",
       "      <td>5</td>\n",
       "      <td>96875986</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000164308</td>\n",
       "      <td>rs35304156</td>\n",
       "      <td>5</td>\n",
       "      <td>96983311</td>\n",
       "      <td>CAT/C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>14.595929;-;3.158018;21.266603;7.549363;6.7724...</td>\n",
       "      <td>212;-;35;363;43;58;-;55;179;186;116;36</td>\n",
       "      <td>4423</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.760042</td>\n",
       "      <td>307.102908</td>\n",
       "      <td>1.909626e-210</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>7.148367e-207</td>\n",
       "      <td>0.000804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000165171</td>\n",
       "      <td>7</td>\n",
       "      <td>73834590</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000165171</td>\n",
       "      <td>rs8629</td>\n",
       "      <td>7</td>\n",
       "      <td>73831189</td>\n",
       "      <td>T/C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>-12.962524;-2.692676;-3.784019;-19.811481;-6.2...</td>\n",
       "      <td>212;32;35;363;43;58;28;55;179;186;116;36</td>\n",
       "      <td>2441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.869423</td>\n",
       "      <td>446.051621</td>\n",
       "      <td>1.474735e-210</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>7.148367e-207</td>\n",
       "      <td>0.001025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000174652</td>\n",
       "      <td>19</td>\n",
       "      <td>9412429</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000174652</td>\n",
       "      <td>rs7258023</td>\n",
       "      <td>19</td>\n",
       "      <td>9415237</td>\n",
       "      <td>G/A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>7.119502;-;-;22.120524;5.970851;4.483163;-;5.3...</td>\n",
       "      <td>212;-;-;363;43;58;-;55;179;186;116;36</td>\n",
       "      <td>5860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.830448</td>\n",
       "      <td>744.484720</td>\n",
       "      <td>8.211917e-175</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>2.049330e-171</td>\n",
       "      <td>0.000920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene  GeneChr   GenePos GeneStrand       GeneSymbol         SNP  \\\n",
       "0  ENSG00000164308        5  96875986          +  ENSG00000164308  rs35304156   \n",
       "1  ENSG00000165171        7  73834590          -  ENSG00000165171      rs8629   \n",
       "2  ENSG00000174652       19   9412429          -  ENSG00000174652   rs7258023   \n",
       "\n",
       "   SNPChr    SNPPos SNPAlleles SNPEffectAllele  ...  \\\n",
       "0       5  96983311      CAT/C               C  ...   \n",
       "1       7  73831189        T/C               C  ...   \n",
       "2      19   9415237        G/A               A  ...   \n",
       "\n",
       "   DatasetZScores.1000IBD.ERP113396.ERP114636.GTEx_Colon_Transverse.GTEx_Small_Intestine.SRP063496.SRP068609.SRP077046.SRP096757.SRP129004.SRP155976.SRP189239.  \\\n",
       "0  14.595929;-;3.158018;21.266603;7.549363;6.7724...                                                                                                              \n",
       "1  -12.962524;-2.692676;-3.784019;-19.811481;-6.2...                                                                                                              \n",
       "2  7.119502;-;-;22.120524;5.970851;4.483163;-;5.3...                                                                                                              \n",
       "\n",
       "  DatasetSampleSizes.1000IBD.ERP113396.ERP114636.GTEx_Colon_Transverse.GTEx_Small_Intestine.SRP063496.SRP068609.SRP077046.SRP096757.SRP129004.SRP155976.SRP189239.  \\\n",
       "0             212;-;35;363;43;58;-;55;179;186;116;36                                                                                                                 \n",
       "1           212;32;35;363;43;58;28;55;179;186;116;36                                                                                                                 \n",
       "2              212;-;-;363;43;58;-;55;179;186;116;36                                                                                                                 \n",
       "\n",
       "   NrTestedSNPs  ProportionBetterPermPvals  BetaDistAlpha  BetaDistBeta  \\\n",
       "0          4423                        0.0       0.760042    307.102908   \n",
       "1          2441                        0.0       0.869423    446.051621   \n",
       "2          5860                        0.0       0.830448    744.484720   \n",
       "\n",
       "   BetaAdjustedMetaP  PvalueNominalThreshold           qval   varbeta  \n",
       "0      1.909626e-210                0.000060  7.148367e-207  0.000804  \n",
       "1      1.474735e-210                0.000071  7.148367e-207  0.001025  \n",
       "2      8.211917e-175                0.000036  2.049330e-171  0.000920  \n",
       "\n",
       "[3 rows x 30 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df[\"varbeta\"] = sum_df[\"MetaSE\"] ** 2\n",
    "sum_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome', 'base_pair_location', 'effect_allele', 'other_allele',\n",
       "       'beta', 'standard_error', 'effect_allele_frequency', 'p_value',\n",
       "       'variant_id', 'rsid', 'FreqSE', 'MinFreq', 'MaxFreq', 'Direction',\n",
       "       'HetISq', 'HetChiSq', 'HetDf', 'HetPVal', 'hm_coordinate_conversion',\n",
       "       'hm_code', 'varbeta'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export adjusted sum stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df.to_csv(\"/groups/umcg-fg/tmp04/projects/gut-bulk/ongoing/2024-02-07-GutPublicRNASeq/extra_scripts/coloc/gwas.tsv\", sep=\"\\t\", header=True, index=False)\n",
    "sum_df.to_csv(\"/groups/umcg-fg/tmp04/projects/gut-bulk/ongoing/2024-02-07-GutPublicRNASeq/extra_scripts/coloc/sum.tsv\", sep=\"\\t\", header=True, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
